### PR TITLE
Advanced TM validation Part 1

### PR DESCRIPTION
**Pull request by fatadel**
Relates to #364

### CODE REVIEW

File: packages/core/shared.js [hunk 1]
```diff
@@ -940,4 +942,93 @@ function checkTmOptionalPointer(td){
...
+    const isSubset = (arr1, arr2) => arr2.every((e) => arr1.includes(e));
+
+    for (const affordanceType of ['properties', 'actions', 'events']) {
+        // Combine all and optional into one => required
+        const required = tmAffordances[affordanceType].all.filter(
+            e => !tmAffordances[affordanceType].optional.includes(e));
+
+        if (!isSubset(Object.keys(td[affordanceType]), required)) {
+            // Some required fields are missing
+            return false;
```
Comment (egekorkan) [inline packages/core/shared.js:987]: should be report as well

**Review by egekorkan — CHANGES_REQUESTED** [review]
There are two assertions to check for tm:optional:
- thing-model-td-generation-processor-type : All required interactions (not listed in tm:optional) MUST be taken over to the Partial TD instance. 
  - This should pass if there is no tm:optional as well.
  - This cannot be not-impl
- thing-model-td-generation-processor-optional : All optional interactions (listed in tm:optional) MAY be taken over to the Partial TD instance. 
  - At least one optional affordance in the TD means this is a pass.
  - If TD does not have any of the optional, this is a not-impl
  - This cannot fail

File: packages/cli/index.js [hunk 2]
```diff
@@ -510,7 +511,11 @@ function coreValidation(input) {
  */
 function checkTd(td, suite) {
 
-    tdValidator(td, console.log, { checkDefaults: myArguments.defaults, checkJsonLd: myArguments.jsonld }, suite)
+    tdValidator(td, console.log, {
```
[the remark this answers is not in the document]
Comment (fatadel) [reply packages/cli/index.js:514]: No, since this is the only statement inside `checkTd` function, plus there is already `then`.

**Comment by fatadel** [comment]
> There are two assertions to check for tm:optional:
> 
> * thing-model-td-generation-processor-type : All required interactions (not listed in tm:optional) MUST be taken over to the Partial TD instance.
>   
>   * This should pass if there is no tm:optional as well.
>   * This cannot be not-impl
> * thing-model-td-generation-processor-optional : All optional interactions (listed in tm:optional) MAY be taken over to the Partial TD instance.
>   
>   * At least one optional affordance in the TD means this is a pass.
>   * If TD does not have any of the optional, this is a not-impl
>   * This cannot fail

I guess the first one can be `not-impl` when td does not link to tm, right?

**Comment by egekorkan** [comment]
> I guess the first one can be not-impl when td does not link to tm, right?

That's actually correct

**Comment by egekorkan** [comment]
Just tried the Netlify preview and there is an error in the console.log:

> Uncaught (in promise) TypeError: can't convert undefined to object
    checkLinkedAffordances https://deploy-preview-383--playground-testing.netlify.app/node_modules/@thing-description-playground/core/dist/web-bundle.min.js:59
web-bundle.min.js:59:160242

The bigger problem is that, even with this error, the TM based validation succeeds. I will try on my local machine tomorrow.

File: packages/core/shared.js [hunk 3]
```diff
@@ -957,4 +960,178 @@ function checkTmOptionalPointer(td){
...
+
+    const results = [];
+    const requiredAssertion = false;
+    const optionalAssertion = false;
+    for (const affordanceType of ['properties', 'actions', 'events']) {
+        // Combine all and optional into one => required
+        const required = tmAffordances[affordanceType].all.filter(
+            e => !tmAffordances[affordanceType].optional.includes(e));
+
+        if (!isSubset(Object.keys(td[affordanceType]), required) && !requiredAssertion) {
```
Comment (egekorkan) [inline packages/core/shared.js:1085]: Here I get:

       

>  if (!isSubset(Object.keys(td[affordanceType]), required) && !requiredAssertion) {
>                              ^
> 
> TypeError: Cannot convert undefined or null to object
>     at Function.keys (<anonymous>)
>     at Object.checkLinkedAffordances (/home/eko/git/thingweb-playground/packages/core/shared.js:1087:30)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async /home/eko/git/thingweb-playground/packages/core/index.js:178:59
> 
> Node.js v19.2.0

The object keys actual get logged correctly when I put add console.log statements so I am not sure where the error comes from

**Review by egekorkan — CHANGES_REQUESTED** [review]

File: packages/core/shared.js [hunk 4]
```diff
@@ -957,4 +960,178 @@ function checkTmOptionalPointer(td){
...
+
+    // Check if arr2 is subset of arr1,
+    // i.e., all elements of arr2 are contained in arr1
+    const isSubset = (arr1, arr2) => arr2.every(e => arr1.includes(e));
+
+    // Check if arr1 and arr2 have any intersection
+    const haveIntersection = (arr1, arr2) => arr1.some(e => arr2.includes(e));
+
+    const results = [];
+    const requiredAssertion = false;
```
Comment (egekorkan) [inline packages/core/shared.js:1078]: these cannot be const as well, they get reassigned

**Comment by egekorkan** [comment]
There is something not working. [This TD](https://deploy-preview-383--playground-testing.netlify.app/#C4EwVgzg9gdg3gKAAQqQIgJYjQLnQVwCcYcIMBbABwBsBTNAGmVTQAEBjWYWgD2F3QALYMEoQcAegkB3WQDppAZjlRCAcwkAmAAybNMqMAmgJANwCMc842Yo0wDMDoC0AWQCeAGQCGVACqCGDBqNqjoILQQ7IQYlA6wLgBq3tRYSH4AIkiclBi0IEgAZoRQ5EjAgrRIEJS07ADkEEUYhBDASLy+NPRMYWgQdUSO7hm0hUGOGLAQAohhfQBG3mTsAPpRs7bzfVGV5PR4aEsrodvzmDAuld4RhGhbYQC+W4+9LAPsQ8DuAgDaDyxjhg1hstgBdN52SglWqEByRTZndBtbzAfAzPBzJEsb61FxtGLBU7Y9CFVTkDFIf4k7ZYmlnNCCQhjK4iMSSCTkdzULpyTpUOhyTjkCQotEzSH085QSh-AFS87Mm7QmW0OE-SUK7ZoaQxbgq2Hfe5a+YQ+VStCcGDca1+dx4w7eSg0YGoqYwCSQBLmpHPelm+loKALAaEUzeBbOPCFFIDTUMpUgADyMGoP2jsdo8e1usctBTaYEMeoA3NftQry2aG87HiMEpdPOwCgajUUaQjaRaDJhApcpNna1jOZhVZonEUi5PKofJ4XUFwuMLbbPR9NKDssOQVMUAA1rQa3XiSb0FabcA7Q70E6Xew3bBPdBLmvtuWSQH1xBvIUDkVM9nziwWgqEMWhrSLTMyxeSU0FSGBd0pakzkHbVmWoFxcVXQMmRZQ5hHHDk1EcQR8AWIVSmMQJgmkWgFkooI1BohYAFoaG8dw1BKfAYBACRI2DTllm4QgJEoGtd28NRIgkThmQkfluggYwKTMFIsAkGU6xSOQvUuAC+kwlwb1SO862UgBqXTjTON8kDBBBHiAA) links to the TM at https://github.com/thingweb/thingweb-playground/blob/master/packages/core/examples/tms/valid/optional.json where the event is not optional. However, the validation still passes

**Comment by egekorkan** [comment]
I had to fix some of the examples and the corresponding expected test results. It is looking good now! :)

**Pull request merged**